DQMServices/Core/src/DQMStore.cc

```diff
@@ -2126,44 +2130,79 @@ DQMStore::reset(void)
 /** Invoke this method after flushing all recently changed monitoring.
     Clears updated flag on all MEs and calls their Reset() method. */
 void
-DQMStore::forceReset(void)
+DQMStore::forceReset()
 {
-  MEMap::iterator mi = data_.begin();
-  MEMap::iterator me = data_.end();
+  auto mi = data_.begin();
+  auto me = data_.end();
   for ( ; mi != me; ++mi)
   {
     if (forceResetOnBeginLumi_ && ((*mi).getLumiFlag() == false))
       continue;
-    MonitorElement &me = const_cast<MonitorElement &>(*mi);
+    auto &me = const_cast<MonitorElement &>(*mi);
     me.Reset();
     me.resetUpdate();
   }
 
   reset_ = true;
 }
 
+
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////
+/** Called after all globalBeginLuminosityBlock.
+ * Reset and mark global per-lumi MEs (or all MEs if LSbasedMode) with
+ * the current lumisection, so they can be reused and properly saved.
+ */
+void
+DQMStore::postGlobalBeginLumi(const edm::GlobalContext &gc)
+{
+  static const std::string null_str("");
+
+  auto const& lumiblock = gc.luminosityBlockID();
+  auto run  = lumiblock.run();
+  auto lumi = lumiblock.luminosityBlock();
+
+  const MonitorElement begin(&null_str, null_str, run, 0, 0);
+  const MonitorElement end(&null_str, null_str, run + 1, 0, 0);
+  auto i = data_.lower_bound(begin);
+  const auto e = data_.lower_bound(end);
+  while (i != e) {
+    auto& me = const_cast<MonitorElement&>(*i++);
+    // skip per-stream MEs
+    if (me.data_.moduleId != 0)
+      continue;
+    // skip per-run MEs
+    if (not LSbasedMode_ and not me.getLumiFlag())
+      continue;
+    me.Reset();
+    me.resetUpdate();
+    me.setLumi(lumi);
```

Comment:
this definitely cannot work, at it changes the *key* of the `set` that holds all MEs